### PR TITLE
[WIP] GPU support

### PR DESCRIPTION
**Pull request by SubhasmitaSw**
**What type of PR is this?**
/kind api-change
//kind feature

**What this PR does / why we need it**:
In support to accommodate the API adjustments needed to manage GPU acceleration of the GCP instances in CAPG.

**Special notes for your reviewer**:
This is a WIP PR, additional controller changes to be added successively. 

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
NONE
```

cc @richardcase @dims @cpanato

### CODE REVIEW

File: controllers/gcpmachine_controller.go
Comment (aniruddha2000) [inline controllers/gcpmachine_controller.go:50]: Can we rename the `AccleratorType` to `AcceleratorConfig`?

File: api/v1beta1/gcpmachine_types.go
Comment (SubhasmitaSw) [reply api/v1beta1/gcpmachine_types.go:133]: yes, as an env var

File: api/v1beta1/gcpmachine_types.go
Comment (SubhasmitaSw) [reply api/v1beta1/gcpmachine_types.go:79]: Both way are fine, 'GCP' prefix makes it more relatable wrt, we should go with `AcceleratorConfig` more standard way.

File: api/v1beta1/gcpmachine_types.go
Comment (richardcase) [inline api/v1beta1/gcpmachine_types.go:78]: We can probably use `// +kubebuilder:default:=1` here.

File: api/v1beta1/gcpmachine_types.go
Comment (richardcase) [inline api/v1beta1/gcpmachine_types.go:112]: Can we add `// +optional` as this is optional.

File: api/v1beta1/gcpmachine_types.go
Comment (richardcase) [inline api/v1beta1/gcpmachine_types.go:121]:
```suggestion
	// +kubebuilder:default:=true
```

**Review by richardcase — CHANGES_REQUESTED** [review]
Great work, lets chat via zoom about the comments.

File: api/v1beta1/gcpmachine_types.go
Comment (aniruddha2000) [reply api/v1beta1/gcpmachine_types.go:83]: As we have a default accelerator type as `nvidia-tesla-t4` does that affect the code while reconciling here in the code base?
Like if we make the default as `nvidia-tesla-k80` here then what will be the difference?

Comment (richardcase) [reply api/v1beta1/gcpmachine_types.go:83]: I would say if as user is choosing to add an accelerator it should be required they say which type instead of defaulting. So marking it required and having no default.

File: api/v1beta1/gcpmachine_types.go
Comment (SubhasmitaSw) [reply api/v1beta1/gcpmachine_types.go:78]: For the current case the validation of accelerators during the instance creation does not allow `int` to be used, further if we change the static validation function to a `conditional check` during reconciliation we can use `int`.

Comment (richardcase) [reply api/v1beta1/gcpmachine_types.go:78]: As a principle, our API doesn't have to match the GGCP exactly, so we have use different data types if needed (as long as we can caste without loosing data).

**Comment by aniruddha2000** [comment]
Added the getters for the new GPU API fields so we can get the items and use them in reconciliation.
@richardcase do we need setters to set those values in case not provided?

File: api/v1beta1/gcpmachine_types.go
Comment (aniruddha2000) [inline api/v1beta1/gcpmachine_types.go:92]: https://pkg.go.dev/google.golang.org/api@v0.86.0/compute/v1#AcceleratorConfig

here the count is int64 should we make it int > int64?

Comment (aniruddha2000) [reply api/v1beta1/gcpmachine_types.go:92]: Otherwise, I guess we might have to convert everything back again while passing these values while reconciliation.

Comment (aniruddha2000) [reply api/v1beta1/gcpmachine_types.go:92]: https://pkg.go.dev/google.golang.org/api@v0.86.0/compute/v1#Scheduling.AutomaticRestart

The AutomaticRestarts accepts a `*bool` should we also change this?

Comment (aniruddha2000) [reply api/v1beta1/gcpmachine_types.go:92]:
```suggestion
	Count int `json:"count,omitempty"`
```

File: cloud/services/compute/instances/reconcile.go
Comment (richardcase) [reply cloud/services/compute/instances/reconcile.go:141]: Also, seeing this i see an instance can have multiple accelerator cards attached. 

@SubhasmitaSw @aniruddha2000 - should we change the GCPMachineSpec to be a slice?

File: api/v1beta1/gcpmachine_types.go
Comment (aniruddha2000) [reply api/v1beta1/gcpmachine_types.go:92]: `int` and `bool` is ok we just have to convert while passing
we should focusing on the actual logic 😃

File: cloud/interfaces.go
Comment (aniruddha2000) [reply cloud/interfaces.go]: Reverted

File: cloud/services/compute/instances/reconcile.go
Comment (aniruddha2000) [reply cloud/services/compute/instances/reconcile.go]: 🤔 we can so that it will satisfy the GCP behavior

File: cloud/scope/machine.go
Comment (aniruddha2000) [inline cloud/scope/machine.go:189]: I think we don't need these changes.

Comment (aniruddha2000) [reply cloud/scope/machine.go:189]: As I already reverted all these in the previous commit

**Comment by cpanato** [comment]
@SubhasmitaSw @aniruddha2000 any status on this?

**Comment by aniruddha2000** [comment]
@cpanato Just a little bit in the documentation is remaining.
I and @SubhasmitaSw are currently facing some difficulty understanding the e2e behavior.

**Comment by richardcase** [comment]
> @cpanato Just a little bit in the documentation is remaining. I and @SubhasmitaSw are currently facing some difficulty understanding the e2e behavior.

@SubhasmitaSw @aniruddha2000  - did you want to chat about the e2e?

File: config/default/manager_image_patch.yaml
Comment (richardcase) [inline config/default/manager_image_patch.yaml:11]: We won't be able to merge this as is....we need to change this back

File: test/e2e/e2e_test.go
Comment (richardcase) [inline test/e2e/e2e_test.go:191]: `GPUCount` has been specified here but its commented out on `VerifyMachineGPUAttachedSpecInput`.

**Review by richardcase — COMMENTED** [review]
Have these tests been run locally to ensure they work as expected?

**Comment by SubhasmitaSw** [comment]
@richardcase  Everything else sets fine except it fails for `make test-e2e-local` giving out error as: 

```
# sigs.k8s.io/cluster-api/internal/topology/variables
../../../../go/pkg/mod/sigs.k8s.io/cluster-api@v1.1.5/internal/topology/variables/clusterclass_variable_validation.go:139:100: not enough arguments in call to structuraldefaulting.ValidateDefaults
        have (*field.Path, *"k8s.io/apiextensions-apiserver/pkg/apiserver/schema".Structural, bool, bool)
        want (context.Context, *field.Path, *"k8s.io/apiextensions-apiserver/pkg/apiserver/schema".Structural, bool, bool)
```

**Comment by richardcase** [comment]
We need to get the images published first. I will check on that first

**Comment by cpanato** [comment]
> We need to get the images published first. I will check on that first

let me check that today

**Comment by FischerLGLN** [comment]
> > We need to get the images published first. I will check on that first
> 
> let me check that today

If the gpu param exist, I can test that with Nvidia GPU Operator, if it works flawlessly.

**Comment by FischerLGLN** [comment]
Hmm, merge conflict in go.sum and go.mod

**Comment by SubhasmitaSw** [comment]
I'll tidy this up!

**Comment by eranco74** [comment]
@SubhasmitaSw 
Any update?

**Comment by richardcase** [comment]
> @SubhasmitaSw Any update?

This is currently blocked waiting for the image-builder changes to merge. I will look at unblocking this asap.

Until image builder changes are merged and images are available, adding an explicit:

/hold

**Comment by elmiko** [comment]
thanks for the update @richardcase , i'm adding myself to the subscribe list here as i a taking a look at compatibility with openshift for the changes.

**Comment by richardcase** [comment]
I am picking this up again after the holiday break.

File: api/v1beta1/gcpmachine_types.go
Comment (JoelSpeed) [inline api/v1beta1/gcpmachine_types.go:136]: As you've created a typed string, this field should be using that typed string
```suggestion
	OnHostMaintenance OnHostMaintenance `json:"onHostMaintenance,omitempty"`
```

Also, and I'm not sure how much this project will agree, Kube API conventions would recommend that the constant values here are `Terminate` and `Migrate`, matching the PascalCase convention for enums, rather than matching the cloud provider values

Not sure what the maintainers prefer but thought I should mention

Comment (richardcase) [reply api/v1beta1/gcpmachine_types.go:136]: Thanks for raising this @JoelSpeed...especially considering the conversation on the confidential compute PR.

Which reminds me i need to get the image-builder PR fixed/merged....

**Comment by richardcase** [comment]
/remove-lifecycle rotten

I will pick up the image building side so that we can get this merged.

**Pull request closed without merging**
